flink-streaming-java/src/main/java/org/apache/flink/streaming/api/datastream/DataStream.java

```diff
@@ -1172,8 +1173,10 @@ public ExecutionConfig getExecutionConfig() {
 	 * @param <R>
 	 *            type of the return stream
 	 * @return the data stream constructed
+	 * @see #transform(String, TypeInformation, OneInputStreamOperatorFactory)
 	 */
 	@PublicEvolving
+	@Deprecated
 	public <R> SingleOutputStreamOperator<R> transform(String operatorName, TypeInformation<R> outTypeInfo, OneInputStreamOperator<T, R> operator) {
```

Comment:
You can eliminate code duplication (this and the newly introduced method) by introducing a private method (with more broader `operatorFactory` argument) like:
```
	private <R> SingleOutputStreamOperator<R> transform(
        String operatorName, TypeInformation<R> outTypeInfo, StreamOperatorFactory<T, R> operatorFactory) {
        ...
    }
```

The old methods would just delegate call to the common implementation, for example:
```
	public <R> SingleOutputStreamOperator<R> transform(String operatorName, TypeInformation<R> outTypeInfo, OneInputStreamOperator<T, R> operator) {
        return transform(operatorName, outTypeInfo, SimpleOperatorFactory.of(operator));
    }
```

Regarding, the deprecation of this api: I think this api method is more user targeted (`OneInputStreamOperator` is `@PublicEvolving`) and it's more or less ok for users to write their own operators as Java classes.

Reply:
Eliminated duplicate and removed @Deprecated.